text/0000-cross-type-number-comparison.md

```diff
@@ -0,0 +1,82 @@
+
+This is an annoyance and even a point of confusion for newer programmers. Additionally, hand-rolled comparisons are prone to subtle mistakes regarding unsigned/signed integers or float rounding. Also, achieving optimal performance is non-trivial and even more prone to mistakes.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Makes it possible to compare numbers of different types to each other, by adding `PartialOrd`/`Ord` and `PartialEq`/`Eq` trait implementations for different number types:
+- `3 < 3.5` is `true`
+- `200_u8 > -1_i8` is `true`
+- `10000000000000001_i64 == 10000000000000000_f64` is `false`
+- ...
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There will be macros to generate the trait implementations for float<->int comparison, int<->int comparison, and float<->float comparison. The actual implementation can be adapted from @orlp's [num-ord crate](https://github.com/orlp/num-ord).
+
+Comparing any integer to a float NaN returns `false` (`None` in `partial_cmp()`)
+
+In a comparison, floats are compared by the exact mathematical value they represent according to IEEE-754. For example:
+- `19999999_f32 == 20000000` because `19999999` is rounded to `20000000` in 32 bit floats
+    - In other words, the number `19999999` _does not exist_ in 32 bit floats
+- `20000000_f32 != 20000001` because both literals represent their value exactly, and the values are different
+- `19999999_f64 != 20000000` because f64 has enough precision to exactly represent `19999999`
+
+Type inference shouldn't be affected by this change, because the generic parameter on `PartialOrd` and `PartialEq` is defaulted to `Self`.
+Therefore, expressions like `1 == 1_u8` will continue to work by correctly inferring the first literal to be a `u8`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Cross-type comparisons are potentially expensive
+    - Unsigned<->signed integer comparisons are barely slower than native same-type comparisons: 5.1ns vs 5.0ns on my machine [playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5a8704dfd953be0d0e750b2d916ccc7f)
+    - However, float<->int comparisons remain to be benchmarked and optimized
+- Cross-type comparisons may hint at a design problem in the code
+    - Quite subjective, and better solved with a lint
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Restructuring the code to use the same number type everywhere
+
+This is not always possible or desirable.
+
+## Using a third-party library like [num-ord](https://github.com/orlp/num-ord)
+
+Requires wrapping number types in the third-party newtype which is ergonomic and not possible in some situations.
```

Comment:
did you mean *unergonomic* instead?

Reply:
what are the "not possible" situations? could you elaborate on that?

Reply:
Good point, I'll address the "not possible" situations in a potential future proposal which proposes ordering functions instead of trait implementations